sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -170,17 +172,22 @@ object ExtractFiltersAndInnerJoins extends PredicateHelper {
   }
 }
 
+
 /**
  * A pattern that collects all adjacent unions and returns their children as a Seq.
  */
 object Unions {
   def unapply(plan: LogicalPlan): Option[Seq[LogicalPlan]] = plan match {
-    case u: Union => Some(collectUnionChildren(u))
+    case u: Union => Some(collectUnionChildren(u :: Nil, Seq.empty[LogicalPlan]))
     case _ => None
   }
 
-  private def collectUnionChildren(plan: LogicalPlan): Seq[LogicalPlan] = plan match {
-    case Union(l, r) => collectUnionChildren(l) ++ collectUnionChildren(r)
-    case other => other :: Nil
+  @tailrec
+  private def collectUnionChildren(
+      plan: List[LogicalPlan],
+      children: Seq[LogicalPlan]): Seq[LogicalPlan] = plan match {
+    case Nil => children
+    case Union(grandchildren) :: res => collectUnionChildren(grandchildren.toList ++ res, children)
+    case other :: res => collectUnionChildren(res, children :+ other)
```

Comment:
Is appending to a list linear in terms of performance? Then this would be a quadratic loop, which can be pretty bad for hundreds of thousands of partitions.